src/store/filestore.rs

```diff
@@ -113,97 +47,256 @@ impl LocalFileStore {
         }
     }
 
-    fn make_annotation_path(&self, class: &str, hash: &str, name: &str, version: &str) -> PathBuf {
+    fn make_dir_path<T>(&self, hash: &str) -> PathBuf {
         PathBuf::from(format!(
-            "{}/{}/{}/{}/{}-{}.yaml",
+            "{}/{}/{}",
             self.directory.to_string_lossy(),
-            "annotation",
-            class,
-            name,
+            get_type_name::<T>(),
             hash,
-            version,
         ))
     }
 
-    fn make_spec_path(&self, class: &str, hash: &str) -> PathBuf {
+    fn make_path<T>(&self, hash: &str, file_name: &str) -> PathBuf {
+        let mut path = self.make_dir_path::<T>(hash);
+        path.push(file_name);
+        path
+    }
+
+    fn make_anno_path<T>(&self, hash: &str, file_name: &str) -> PathBuf {
+        let mut path = self.make_dir_path::<T>(hash);
+        path.push("annotations");
+        path.push(file_name);
+        path
+    }
+
+    // Generic function for save load list delete
+    /// Generic func to save all sorts of item
+    ///
+    /// Example usage inside LocalFileStore
+    /// ``` markdown
+    /// let pod = Pod::new(); // For example doesn't actually work
+    /// self.save_item(pod, &pod.annotation, &pod.hash).unwrap()
+    /// ```
+    fn save_item<T: Serialize>(
+        &self,
+        item: &T,
+        hash: &str,
+        annotation: Option<&Annotation>,
+    ) -> Result<(), Box<dyn Error>> {
+        // Save the item first
+        Self::save_file(
+            &self.make_path::<T>(hash, "spec.yaml"),
+            &to_yaml::<T>(item)?,
+            false,
+        )?;
+
+        // Save the annotation file and throw and error if exist
+        if let Some(value) = annotation {
+            // Annotation exist, thus save it
+            Self::save_file(
+                &self.make_anno_path::<T>(hash, &format!("{}-{}.yaml", value.name, value.version)),
+                &serde_yaml::to_string(value)?,
+                true,
+            )?;
+
+            // Update name-ver index tree
+            self.insert_item_to_name_ver_index::<T>(&value.name, &value.version, hash)?;
+
+            let mut file = fs::File::open(self.make_name_ver_index_path::<T>())?;
+            let mut temp = Vec::new();
+            file.read_to_end(&mut temp)?;
+        }
+
+        Ok(())
+    }
+
+    /// Generic function for loading spec.yaml into memory
+    fn load_item<T: DeserializeOwned>(
```

Comment:
change to 
load_item<T:asjdfklajsdf>(&self, spec; ResourceSpec)

enum ResourceSpec {
hash(string,
annotation(String, String)
}

Reply:
updated